flink-tests/src/test/java/org/apache/flink/test/checkpointing/NotifyCheckpointAbortedITCase.java

```diff
@@ -0,0 +1,576 @@
+	 *                                  |
+	 *                                  |--> BarrierDelayMap -> BeforeExecuteCheckpointSink
+	 */
+	@Test(timeout = TEST_TIMEOUT)
+	public void testNotifyCheckpointAborted() throws Exception {
+		Duration timeout = Duration.ofMillis(TEST_TIMEOUT);
+		Deadline deadline = Deadline.now().plus(timeout);
+
+		final StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+		env.enableCheckpointing(200, CheckpointingMode.EXACTLY_ONCE);
+		env.getCheckpointConfig().enableUnalignedCheckpoints(unalignedCheckpointEnabled);
+		env.getCheckpointConfig().setTolerableCheckpointFailureNumber(1);
+		env.disableOperatorChaining();
+		env.setParallelism(1);
+
+		final StateBackend notifiedStateBackend = new NotifiedStateBackend(checkpointPath);
+		env.setStateBackend(notifiedStateBackend);
+
+		SingleOutputStreamOperator<Integer> normalMapStream = env
+			.addSource(new NormalSource()).name("NormalSource")
+			.keyBy((KeySelector<Tuple2<Integer, Integer>, Integer>) value -> value.f0)
+			.map(new NormalMapFunction()).name("NormalMap");
+
+		normalMapStream.
+			transform(StuckAsyncCheckpointMapName, TypeInformation.of(Integer.class), new StuckAsyncCheckpointMap())
+			.addSink(new DeclineSink()).name("DeclineSink");
+
+		normalMapStream
+			.transform("BarrierDelayMap", TypeInformation.of(Integer.class), new BarrierDelayMap())
+			.transform("BeforeExecuteCheckpointSink", TypeInformation.of(Object.class), new BeforeExecuteCheckpointSink());
+
+		final ClusterClient<?> clusterClient = cluster.getClusterClient();
+		JobGraph jobGraph = env.getStreamGraph().getJobGraph();
+		JobID jobID = jobGraph.getJobID();
+
+		ClientUtils.submitJob(clusterClient, jobGraph);
+
+		TestingCompletedCheckpointStore.addCheckpointLatch.await(deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS);
+
+		Set<File> localStoredStates = new HashSet<>();
+		while (deadline.hasTimeLeft()) {
+			// found the local state manager has stored states for checkpoint-1.
+			localStoredStates.addAll(collectLocalStoredStates());
+			if (!localStoredStates.isEmpty()) {
+				break;
+			}
+		}
+
+		// let the checkpoint-1 failed finally.
+		TestingCompletedCheckpointStore.abortCheckpointLatch.trigger();
+		while (deadline.hasTimeLeft()) {
+			// verify the local state manager has been cleaned up once notified as checkpoint-1 aborted.
+			localStoredStates.removeIf(file -> !file.exists());
+			if (localStoredStates.isEmpty()) {
+				break;
+			}
+		}
+
+		// wait for StuckAsyncCheckpointMap notified as checkpoint aborted.
+		StuckAsyncCheckpointMap.checkpointAbortedLatch.await(deadline.timeLeft().toMillis(), TimeUnit.MILLISECONDS);
+		assertTrue(stuckAsyncSnapshotStrategy.blockingRunnableFuture.isCancelled());
+
+		// verify BeforeExecuteCheckpointSink never execute checkpoint-1.
+		while (deadline.hasTimeLeft()) {
+			if (!BeforeExecuteCheckpointSink.snapshotIds.isEmpty()) {
```

Comment:
Can we use accumulators instead of a static field?
It could save from sync/wait and improve readability (no static field).